bundles/org.openhab.binding.viessmann/src/main/java/org/openhab/binding/viessmann/internal/handler/DeviceHandler.java

```diff
@@ -0,0 +1,1165 @@
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.Bridge;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BridgeHandler;
+import org.openhab.core.thing.binding.ThingHandlerCallback;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.link.ItemChannelLink;
+import org.openhab.core.thing.link.ItemChannelLinkRegistry;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.StateOption;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * The {@link DeviceHandler} is responsible for handling DeviceHandler
+ *
+ *
+ * @author Ronny Grun - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = DeviceHandler.class)
+public class DeviceHandler extends ViessmannThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(DeviceHandler.class);
+
+    private static final Gson GSON = new GsonBuilder().setDateFormat("yyyy-MM-dd HH:mm:ss").create();
+
+    private ThingsConfig config = new ThingsConfig();
+
+    private final Map<String, HeatingCircuit> heatingCircuits = new HashMap<>();
+
+    private final ItemChannelLinkRegistry linkRegistry;
+
+    private final TranslationProvider i18Provider;
+    private final LocaleProvider localeProvider;
+
+    private final Bundle bundle;
+
+    public DeviceHandler(Thing thing, ViessmannDynamicStateDescriptionProvider stateDescriptionProvider,
+            ItemChannelLinkRegistry linkRegistry, TranslationProvider i18Provider, LocaleProvider localeProvider) {
+        super(thing, stateDescriptionProvider);
+        this.linkRegistry = linkRegistry;
+        this.i18Provider = i18Provider;
+        this.localeProvider = localeProvider;
+        bundle = FrameworkUtil.getBundle(getClass());
+    }
+
+    @Override
+    public void initialize() {
+        ThingsConfig config = getConfigAs(ThingsConfig.class);
+        this.config = config;
+        if (config.deviceId.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid device id setting");
+            return;
+        }
+        updateProperty(PROPERTY_ID, config.deviceId); // set representation property used by discovery
+
+        migrateChannelIds();
+
+        initDeviceState();
+        logger.trace("Device handler finished initializing");
+    }
+
+    private void migrateChannelIds() {
```

Comment:
I'm not so familiair with channel migration like this. But when looking at it i wonder if you can somehow set a property like thingtypeVersion that can be recognized to skip the mgiration if it was already done.

Trying to migrate everytime that everything is initialized is costly and not needed. @jlaur do you have any advice?